File: Cryptographic_Frontend/cf/GenerateKeyPairMethod.cpp
#include "GenerateKeyPairMethod.hpp"

#include "Argument.hpp"
#include "ResponseMessage.hpp"
#include <json/json.h>

using namespace cf;
using std::string;

GenerateKeyPairMethod::GenerateKeyPairMethod(string keyType, 
                                             int keySize, 
                                             string publicExponent)
  : Method("GenerateKeyPair")
{
  /* methodMessage es dueño de los argumentos */
  addArgument(ArgumentPtr(new StringArgument("keyType", keyType)));
  addArgument(ArgumentPtr(new IntegerArgument("keySize", keySize)));
  addArgument(ArgumentPtr(new StringArgument("publicExponent", publicExponent)));
}

ResponseMessagePtr GenerateKeyPairMethod::parseResponse(const std::string & message) {
    Json::Value json;
    Json::Reader reader;
    
    reader.parse(message, json);

    ResponseMessagePtr response(new ResponseMessage());
    response->addValue(ArgumentPtr( new StringArgument("handler", json["handler"].asString()) ));

    return response;
}
// kate: indent-mode cstyle; indent-width 4; replace-tabs on;

File: CryptographicBackend/src/main/java/cl/niclabs/cb/backend/methods/GenerateKeyPairMethod.java
package cl.niclabs.cb.backend.methods;

import cl.niclabs.cb.backend.Method;

public interface GenerateKeyPairMethod extends Method {

  public static class ReturnValue {
    public final String keyHandler;

      public ReturnValue(String keyHandler) {
          this.keyHandler = keyHandler;
      }
  }

  public static class Args {
    public final String keyType;
    public final int keySize;
    public final String publicExponent;

      public Args(String keyType, int keySize, String publicExponent) {
          this.keyType = keyType;
          this.keySize = keySize;
          this.publicExponent = publicExponent;
      }
  }

}
